### PR TITLE
Add support for bundles with empty artifact dir

### PR DESCRIPTION
**Pull request by salmanmashayekh**
* Add `ignore_hidden` to `directory_comparison`
* Add several unittests for directory comparison
* Add a new `noop_compound_bundle` bundle without any artifacts to the tests

### CODE REVIEW

File: microcosm_sagemaker/artifact.py [hunk 1]
```diff
@@ -16,6 +16,9 @@ def __init__(self, path: Union[str, Path]):
     def init(self) -> None:
         self.path.mkdir(parents=True, exist_ok=True)
 
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}({str(self.path)})"
```
Comment (pokey) [inline microcosm_sagemaker/artifact.py]: I know I wrote this line, but on seeing it again I think we should wrap the path in quotes so that it’s a more accurate representation

File: microcosm_sagemaker/testing/directory_comparison.py [hunk 2]
```diff
@@ -32,10 +33,12 @@ def directory_comparison(
     actual_paths = sorted([
         subpath.relative_to(actual_dir)
         for subpath in actual_dir.glob('**/*')
+        if not (subpath.name.startswith(".") and ignore_hidden)  # exclude hidden files if ignore_hidden is True
```
Comment (pokey) [inline microcosm_sagemaker/testing/directory_comparison.py]: I’d make an `_is_hidden` function

File: microcosm_sagemaker/testing/directory_comparison.py [hunk 3]
```diff
@@ -19,6 +19,7 @@ def directory_comparison(
     gold_dir: Path,
     actual_dir: Path,
     matchers: Optional[Mapping[Path, ExtractorMatcherPair]] = None,
+    ignore_hidden: bool = True,
 ):
     """
     Recursively checks the contents of `actual_dir` against the expected
```
Comment (pokey) [inline microcosm_sagemaker/testing/directory_comparison.py]: My PR had added a check that the input directories exist because if not, then the test actually silently passes. This is bad because someone could get the paths wrong and think their test passes. Would be worth adding a test for this corner case (pass in nonexistent paths)

File: microcosm_sagemaker/tests/fixtures/directory_comparison.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            "subdir2": Dir({
+                ".keep": File(""),
+            }),
+        }),
+        actual=Dir({
+            "subdir1": Dir({
+                "hi.txt": File("hi"),
+            }),
+            "hello.txt": File("hello"),
+            "subdir2": Dir(),
```
Comment (pokey) [inline microcosm_sagemaker/tests/fixtures/directory_comparison.py]: It would be worth having a doubly-nested directory containing a hidden file

File: microcosm_sagemaker/tests/fixtures/directory_comparison.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+    # Using ignore_hidden=True
+    DirectoryComparisonTest(
+        gold=Dir({
+            "subdir1": Dir({
+                "hi.txt": File("hi"),
+            }),
+            "hello.txt": File("hello"),
+            "subdir2": Dir({
+                ".keep": File(""),
```
Comment (pokey) [inline microcosm_sagemaker/tests/fixtures/directory_comparison.py]: Could just do no-args `File()`

File: microcosm_sagemaker/artifact.py (hunk 1 above)
Comment (salmanmashayekh) [reply microcosm_sagemaker/artifact.py]: like this @pokey ?
`f'{type(self).__name__}("{str(self.path)}")'`

File: microcosm_sagemaker/testing/directory_comparison.py (hunk 3 above)
Comment (salmanmashayekh) [reply microcosm_sagemaker/testing/directory_comparison.py]: Good catch! I added `test_nonexistent_directory_comparison` to `test_directory_comparison.py`.

File: microcosm_sagemaker/testing/bundle.py [hunk 6]
```diff
@@ -128,8 +128,6 @@ def test_save(self) -> None:
 
 
 class BundleLoadTestCase(BundleTestCase, BundlePredictionChecker):
-    gold_bundle_output_artifact_path: Path
-
```
Comment (pokey) [inline microcosm_sagemaker/testing/bundle.py]: This wasn't necessary; I had removed in my original PR

File: microcosm_sagemaker/testing/directory_comparison.py [hunk 7]
```diff
@@ -11,11 +11,11 @@
 from microcosm_sagemaker.testing.bytes_extractor import ExtractorMatcherPair
 
 
-def identity(x):
+def _identity(x):
     return x
 
 
-def is_hidden(path: Path) -> bool:
+def _is_hidden(path: Path) -> bool:
```
Comment (pokey) [inline microcosm_sagemaker/testing/directory_comparison.py]: Until we have use for these in other files, better to mark them as ["hidden"](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles)

File: microcosm_sagemaker/testing/directory_comparison.py [hunk 8]
```diff
@@ -31,21 +31,27 @@ def directory_comparison(
...
 
+    By default, this function ignores hidden files.  This functionality is
+    useful when you expect an empty directory, because git won't allow checking
+    in an empty directory.  In this situation you can add an empty `.keep` file
+    to the directory to make sure it is checked in.
+
     """
-    matchers = matchers or dict()
+    if matchers is None:
+        matchers = dict()
```
Comment (pokey) [inline microcosm_sagemaker/testing/directory_comparison.py]: The above is more compact, but this is the [recommended](https://docs.python.org/3/faq/programming.html#why-are-default-values-shared-between-objects) idiom, due to the fact that the user could pass in an empty custom dict type

File: microcosm_sagemaker/testing/directory_comparison.py [hunk 9]
```diff
@@ -31,21 +31,27 @@ def directory_comparison(
...
 
     assert_that(gold_dir.exists(), is_(True))
     assert_that(actual_dir.exists(), is_(True))
 
     actual_paths = sorted([
         subpath.relative_to(actual_dir)
         for subpath in actual_dir.glob('**/*')
-        if not (is_hidden(subpath) and ignore_hidden)  # exclude hidden files if ignore_hidden is True
+        if not (ignore_hidden and _is_hidden(subpath))  # exclude hidden files if ignore_hidden is True
```
Comment (pokey) [inline microcosm_sagemaker/testing/directory_comparison.py]: I'd put `ignore_hidden` first because

- it is a bit more readable
- it is a bit more efficient when `ignore_hidden` is `False`, due to short-circuiting.  Not that that really matters at all here but no reason to [prematurely pessimize](https://www.oreilly.com/library/view/c-coding-standards/0321113586/ch10.html) 🤓

File: microcosm_sagemaker/tests/bundles/noop_compound.py [hunk 10]
```diff
@@ -12,6 +12,7 @@
 class NoopCompoundBundle(Bundle):
     """
     This doesn't do any of its own training.
+
```
Comment (pokey) [inline microcosm_sagemaker/tests/bundles/noop_compound.py]: We always put a blank line before the end of docstrings

File: microcosm_sagemaker/tests/directory_test_description.py [hunk 11]
```diff
@@ -13,28 +15,20 @@ def instantiate(self, path: Path) -> None:
         ...
 
 
+@dataclass
 class Dir(Entry):
-    def __init__(self, entries: Optional[Mapping[str, Entry]] = None):
-        if entries is None:
-            entries = dict()
-        self.entries = entries
+    entries: Mapping[str, Entry] = field(default_factory=dict)
```
Comment (pokey) [inline microcosm_sagemaker/tests/directory_test_description.py]: Had the thought that a `dataclass` was cleaner here because it is a semantically richer way to capture all the behaviour I was defining manually

File: microcosm_sagemaker/tests/fixtures/directory_comparison.py [hunk 12]
```diff
@@ -215,7 +214,6 @@ class DirectoryComparisonTest(NamedTuple):
         }),
         actual=Dir({
             "subdir1": Dir({
-                "hi.txt": File("hi"),
```
Comment (pokey) [inline microcosm_sagemaker/tests/fixtures/directory_comparison.py]: I wanted a doubly-nested totally empty dir, because that actually would have broken my version

File: microcosm_sagemaker/testing/directory_comparison.py (hunk 9 above)
Comment (salmanmashayekh) [reply microcosm_sagemaker/testing/directory_comparison.py]: Interesting! I did not know Python is smart enough to stop if the first condition of an AND is False! :D
